cluster/swarm/cluster.go

```diff
@@ -66,35 +66,16 @@ func (c *Cluster) Handle(e *cluster.Event) error {
 
 // CreateContainer aka schedule a brand new container into the cluster.
 func (c *Cluster) CreateContainer(config *dockerclient.ContainerConfig, name string) (*cluster.Container, error) {
+	c.scheduler.Lock()
+	defer c.scheduler.Unlock()
 
-	c.RLock()
-	defer c.RUnlock()
-retry:
-	// FIXME: to prevent a race, we check again after the pull if the node can still handle
-	// the container. We should store the state in the store before pulling and use this to check
-	// all the other container create, but, as we don't have a proper store yet, this temporary solution
-	// was chosen.
 	n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), config)
 	if err != nil {
 		return nil, err
 	}
 
 	if nn, ok := n.(*node); ok {
-		container, err := nn.create(config, name, false)
```

Comment:
I'm okay if this is a temporary fix . But don't you think that locking create container method will increase the overall container launch time ? . If there are 10 create container calls involving 10 different images simultaneously this is a worst case scenario . A better fix would be maintaining a queue of scheduled containers in node struct 
```
lock()
n, err := c.scheduler.SelectNodeForContainer(c.listNodes(), config)
 	if err != nil {
 		return nil, err
 	}
 n.addtoqueue(config)
unlock()
container, err := nn.create(config, name, false)

node.create method {
defer n.removefromqueue(config)
...........
.........
}
```
while applying filters or placing the container according to the scheduling strategy  Add a check for scheduleQueue
This would decrease the overall create container time significantly because we are not waiting for pul Image time